In [1]:
import pandas as pd

## 1.  Attendance Data
    


    Read the data from the `attendance` table and calculate an attendance percentage for each student.  
    One half day is worth 50% of a full day, and 10 tardies is equal to one absence.
    
    

In [2]:
attendance = pd.read_csv('./untidy-data/attendance.csv')
attendance.head()

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [7]:
attendance.shape

(4, 9)

In [14]:
attendance.rename(columns={'Unnamed: 0':'student'}, inplace=True)


In [15]:
attendance.columns

Index(['student', '2018-01-01', '2018-01-02', '2018-01-03', '2018-01-04',
       '2018-01-05', '2018-01-06', '2018-01-07', '2018-01-08'],
      dtype='object')

In [17]:

df_melted = attendance.melt(id_vars= 'student', var_name='day', value_name='attended')
df_melted.head()

student         day attended
0   Sally  2018-01-01        P
1    Jane  2018-01-01        A
2   Billy  2018-01-01        A
3    John  2018-01-01        P
4   Sally  2018-01-02        T

In [20]:
df_melted['grade'] = df_melted.attended.replace({  'P': 1,
                                                   'H': .5,
                                                   'A': 0,
                                                   'T': .9})
df_melted.head()

student         day attended  grade
0   Sally  2018-01-01        P    1.0
1    Jane  2018-01-01        A    0.0
2   Billy  2018-01-01        A    0.0
3    John  2018-01-01        P    1.0
4   Sally  2018-01-02        T    0.9

In [22]:
df_melted.groupby('student').grade.mean()

student
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: grade, dtype: float64

You should end up with something like this:
    
    `name Billy    0.5250 Jane     0.6875 John     0.9125 Sally    0.7625 Name: grade, dtype: float64`
    


## 2.  Coffee Levels
    


    1.  Read the `coffee_levels` table.


In [24]:
df = pd.read_csv('./untidy-data/coffee_levels.csv')
df

hour coffee_carafe  coffee_amount
0      8             x       0.816164
1      9             x       0.451018
2     10             x       0.843279
3     11             x       0.335533
4     12             x       0.898291
5     13             x       0.310711
6     14             x       0.507288
7     15             x       0.215043
8     16             x       0.183891
9     17             x       0.391560
10     8             y       0.189297
11     9             y       0.521502
12    10             y       0.023163
13    11             y       0.235529
14    12             y       0.017009
15    13             y       0.997464
16    14             y       0.058361
17    15             y       0.144644
18    16             y       0.544676
19    17             y       0.594126
20     8             z       0.999264
21     9             z       0.915990
22    10             z       0.144928
23    11             z       0.311495
24    12             z       0.771947
25    13             z       0.398520
26    14             z       0.864464
27    15             z       0.436364
28    16             z       0.280621
29    17             z       0.436677

    2.  Transform the data so that each carafe is in it's own column.


In [26]:
df_pivot = df.pivot_table(index= ['hour'],
                                    columns= 'coffee_carafe',
                                    values= 'coffee_amount',
                                    aggfunc= 'mean').reset_index()
df_pivot

coffee_carafe  hour         x         y         z
0                 8  0.816164  0.189297  0.999264
1                 9  0.451018  0.521502  0.915990
2                10  0.843279  0.023163  0.144928
3                11  0.335533  0.235529  0.311495
4                12  0.898291  0.017009  0.771947
5                13  0.310711  0.997464  0.398520
6                14  0.507288  0.058361  0.864464
7                15  0.215043  0.144644  0.436364
8                16  0.183891  0.544676  0.280621
9                17  0.391560  0.594126  0.436677

In [29]:
df_pivot.columns.name = ''
df_pivot

hour         x         y         z
0     8  0.816164  0.189297  0.999264
1     9  0.451018  0.521502  0.915990
2    10  0.843279  0.023163  0.144928
3    11  0.335533  0.235529  0.311495
4    12  0.898291  0.017009  0.771947
5    13  0.310711  0.997464  0.398520
6    14  0.507288  0.058361  0.864464
7    15  0.215043  0.144644  0.436364
8    16  0.183891  0.544676  0.280621
9    17  0.391560  0.594126  0.436677

    3.  Is this the best shape for the data?


In [ ]:
## It depends on what an observation is.
## However I think this table looks correct since each column seems to be a part of a coordinate.

## 3.  Cake Recipes
    


    1.  Read the `cake_recipes` table. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.


In [37]:
df = pd.read_csv('./untidy-data/cake_recipes.csv')
df.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [38]:
df.shape

(8, 5)

    2.  Tidy the data as necessary.


In [39]:
df[['recipe','position']] = df['recipe:position'].str.split(':', expand=True)
df.head()

recipe:position        225        250        275        300 recipe position
0        a:bottom  61.738655  53.912627  74.414730  98.786784      a   bottom
1           a:top  51.709751  52.009735  68.576858  50.228470      a      top
2        b:bottom  57.095320  61.904369  61.196980  99.248541      b   bottom
3           b:top  82.455004  95.224151  98.594881  58.169349      b      top
4        c:bottom  96.470207  52.001358  92.893227  65.473084      c   bottom

In [40]:
df = df.drop(columns='recipe:position')
df.head()

225        250        275        300 recipe position
0  61.738655  53.912627  74.414730  98.786784      a   bottom
1  51.709751  52.009735  68.576858  50.228470      a      top
2  57.095320  61.904369  61.196980  99.248541      b   bottom
3  82.455004  95.224151  98.594881  58.169349      b      top
4  96.470207  52.001358  92.893227  65.473084      c   bottom

In [42]:
df_tidy = df.melt(
        id_vars=['recipe','position'],
        var_name='temp',
        value_name='tastiness'
        )
df_tidy.head()

recipe position temp  tastiness
0      a   bottom  225  61.738655
1      a      top  225  51.709751
2      b   bottom  225  57.095320
3      b      top  225  82.455004
4      c   bottom  225  96.470207

In [43]:
df_tidy.shape

(32, 4)

    3.  Which recipe, on average, is the best? recipe b


In [45]:
df_tidy.groupby('recipe').tastiness.mean().nlargest(1)

recipe
b    76.736074
Name: tastiness, dtype: float64

    4.  Which oven temperature, on average, produces the best results? 275


In [46]:
df_tidy.groupby('temp').tastiness.mean().nlargest(1)

temp
275    74.886754
Name: tastiness, dtype: float64

    5.  Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees


In [58]:
row = df_tidy.tastiness.idxmax()
type(row)

int

In [66]:
df_tidy.iloc[[row]]

recipe position temp  tastiness
26      b   bottom  300  99.248541


## 4.  **Bonus**: 
explore the other tables in the `tidy_data` database and reshape them as necessary so that they are in a tidy format.